source/reduce/operand_to_const_reduction_pass.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "operand_to_const_reduction_pass.h"
+#include "change_operand_reduction_opportunity.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace reduce {
+
+using namespace opt;
+
+std::vector<std::unique_ptr<ReductionOpportunity>>
+OperandToConstReductionPass::GetAvailableOpportunities(
+    opt::IRContext* context) const {
+  std::vector<std::unique_ptr<ReductionOpportunity>> result;
+  assert(result.empty());
+  for (const auto& constant : context->GetConstants()) {
```

Comment:
It might be clearer to make this the innermost loop.

Reply:
That's actually how I structured it originally, because that would make reduction opportunities for the same operand adjacent to one another.  If two reduction opportunities refer to the same operand, they can't both be applied (because one disables the other).  The reducer takes a greedy approach, as per delta debugging, where blocks of adjacent reduction opportunities are tried, with the block size exponentially shrinking.  Thus it's best to structure things so that adjacent opportunities are likely to be compatible, hence this loop structure.  I'll reflect this in the code comments.

Reply:
Resolving this by adding an explanatory comment in the code.